plugins/pluginmanagers/lazy.nix

```diff
@@ -1,222 +1,264 @@
-            name = "${lib.getName p}";
-            path = p;
+            dir = lib.mkDefault "${plugin}";
+            name = lib.mkDefault "${lib.getName plugin}";
           }
+        else if lib.isString plugin then
+          { __unkeyed = lib.mkDefault plugin; }
         else
-          {
-            name = "${lib.getName p.pkg}";
-            path = p.pkg;
-          };
-      processDependencies =
-        if plugin ? dependencies && plugin.dependencies != null then
-          builtins.concatMap processPlugin plugin.dependencies
-        else
-          [ ];
-    in
-    [ (mkEntryFromDrv plugin) ] ++ processDependencies;
-
-  processedPlugins = builtins.concatLists (builtins.map processPlugin lazyPlugins);
-  lazyPath = pkgs.linkFarm "lazy-plugins" processedPlugins;
-in
-{
-  options = {
-    plugins.lazy = {
-      enable = mkEnableOption "lazy.nvim";
+          plugin;
 
-      gitPackage = lib.mkPackageOption pkgs "git" {
-        nullable = true;
-      };
+      lazyPluginCoercibleType = with types; either str package;
 
-      plugins =
+      lazyPluginType =
         with types;
-        let
-          pluginType = either package (submodule {
-            options = {
-              dir = helpers.mkNullOrOption str "A directory pointing to a local plugin";
-
-              pkg = mkOption {
-                type = package;
-                description = "Vim plugin to install";
-              };
+        types.coercedTo lazyPluginCoercibleType coerceToLazyPluginSpec (
+          submodule (
+            { config, ... }:
+
+            let
+              cfg = config;
+            in
+            {
+              freeformType = attrsOf anything;
+              options = {
+                __unkeyed = mkNullOrOption str ''
+                  The "__unkeyed" attribute can either be one of:
+
+                    - a local plugin directory path.
+                    - a full plugin url.
+                    - a short plugin url.
+                    - a custom name for a plugin.
+
+                  If a short plugin url is given e.g. "echasnovski/mini.ai" it
+                  will be expanded using `plugins.lazy.settings.git.url_format`
+
+                  If a custom name for a plugin is given, the plugin must be
+                  defined somewhere else in the list of plugins. For example:
+
+                  ```nix
+                  plugins.lazy.plugins = with pkgs.vimPlugins; [
+                    # Custom name for vim plugin
+                    "oil"
+
+                    # The plugin gets defined by using the custom name that was
+                    # previously given.
+                    {
+                      name = "oil";
+                      pkg = oil-nvim;
+                    }
+                  ];
+                  ```
+                '';
+
+                dir = mkNullOrOption str ''
+                  A directory pointing to a local plugin path e.g. "~/plugins/trouble.nvim".
+                '';
```

Comment:
I'm confused. Both `dir` and `__unkeyed` say they can be the path to a directory where the plugin is found... Are they the same thing?

Reply:
Ah I see, you had `url` `pkg` and `dir` all mapping internally to `__unkeyed` if defined?

In this case I think we just remove the `__unkeyed` option from this submodule and do that kinda conversion later, when defining `settings.spec`.

Reply:
Gotcha. So these options all map to upstream options, but there's some redundancy?

I think we need to better understand how upstream handles these before finalizing a design here.

As for `__unkeyed` I feel like it can be named something better if this is an option we're intentionally doing transformations to, since our usual rules of "pass exactly what the user configured in `settings` to `setup`" don't really apply here. :thinking: